src/community/features-autopopulate/features-autopopulate-core/src/main/java/org/geoserver/autopopulate/AutopopulateTransactionCallback.java

```diff
@@ -185,37 +187,41 @@ public TransactionRequest beforeTransaction(TransactionRequest request) throws W
                         getFeatureTypeInfo(new NameImpl(element.getTypeName()));
                 try {
                     Update updateElement = (Update) element;
-                    SimpleFeature feature =
-                            DataUtilities.first(getTransactionSource(updateElement).getFeatures());
-                    LOGGER.info("Updating feature: " + feature);
-                    SimpleFeature transformed = applyTemplate(feature);
-                    List<Property> properties = updateElement.getUpdateProperties();
-                    for (org.geotools.api.feature.Property p : transformed.getProperties()) {
-                        if (properties.stream().anyMatch(prop -> match(p, prop))) {
-                            properties.stream()
-                                    .filter(prop -> match(p, prop))
-                                    .forEach(prop -> prop.setValue(p.getValue()));
-                        } else {
-                            Property updateProperty = updateElement.createProperty();
-                            updateProperty.setName(
-                                    new QName(
-                                            featureTypeInfo.getNamespace().getURI(),
-                                            p.getName().getLocalPart()));
-                            updateProperty.setValue(p.getValue());
-                            properties.add(updateProperty);
+                    SimpleFeatureIterator featureIterator =
+                            getTransactionSource(updateElement).features();
```

Comment:
So this is now scrolling through all of the features being updated. But in the end, the update does not depend on the feature attributes, it's just based on env variables. So why go over potentially million of features being updated to compute the target values? I would just consider the target schema instead, should provide all you need?

Reply:
So the code now works like this... we need to update all the UpdateElements properties anyway. It looks like it is iterataing on all the features, but it just uses the features related to the current UpdateElement.
The `getTransactionSource` method applies the `UpdateElement.getFilter` and returns back the iterator.

Reply:
I removed the loop btw since it was not required of course.

Reply:
Good catch! Thank you again for checking it. Updated accordingly.